internal/extsvc/bitbucketserver/client.go

```diff
@@ -64,29 +65,78 @@ type Client struct {
-	// server. https://bitbucket.example.com/plugins/servlet/access-tokens/manage
-	Token string
-
-	// The username and password credentials for accessing the server. Typically these are only
-	// used when the server doesn't support personal access tokens (such as Bitbucket Server
-	// version 5.4 and older). If both Token and Username/Password are specified, Token is used.
-	Username, Password string
+	// Auth is the authentication method used when accessing the server.
+	// Supported types are:
+	// * auth.OAuthBearerToken for a personal access token; see also
+	//   https://bitbucket.example.com/plugins/servlet/access-tokens/manage
+	// * auth.BasicAuth for a username and password combination. Typically
+	//   these are only used when the server doesn't support personal access
+	//   tokens (such as Bitbucket Server 5.4 and older).
+	// * SudoableClient for an OAuth 1 client used to authenticate requests.
+	//   This is generally set using SetOAuth.
+	Auth auth.Authenticator
 
 	// RateLimit is the self-imposed rate limiter (since Bitbucket does not have a concept
 	// of rate limiting in HTTP response headers).
 	RateLimit *rate.Limiter
-
-	// OAuth client used to authenticate requests, if set via SetOAuth.
-	// Takes precedence over Token and Username / Password authentication.
-	Oauth *oauth.Client
 }
 
 // NewClient returns an authenticated Bitbucket Server API client with
 // the provided configuration. If a nil httpClient is provided, http.DefaultClient
 // will be used.
-func NewClient(c *schema.BitbucketServerConnection, httpClient httpcli.Doer) (*Client, error) {
-	u, err := url.Parse(c.Url)
+func NewClient(config *schema.BitbucketServerConnection, httpClient httpcli.Doer) (*Client, error) {
+	client, err := newClient(config, httpClient)
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Authorization == nil {
+		if config.Token != "" {
+			client.Auth = &auth.OAuthBearerToken{Token: config.Token}
+		} else {
+			client.Auth = &auth.BasicAuth{
+				Username: config.Username,
+				Password: config.Password,
+			}
+		}
+	} else {
+		err := client.SetOAuth(
+			config.Authorization.Oauth.ConsumerKey,
+			config.Authorization.Oauth.SigningKey,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "authorization.oauth.signingKey")
+		}
+	}
+
+	return client, nil
+}
+
+// NewClientWithAuthenticator returns an authenticated Bitbucket Server API
+// client with the provided configuration, but using the given Authenticator and
+// ignoring any authentication related fields in the configuration. If a nil
+// httpClient is provided, http.DefaultClient will be used.
+func NewClientWithAuthenticator(config *schema.BitbucketServerConnection, httpClient httpcli.Doer, a auth.Authenticator) (*Client, error) {
+	switch a.(type) {
+	case *auth.OAuthBearerToken, *auth.BasicAuth, *SudoableOAuthClient:
+		// Excellent.
+	case nil:
+		return nil, errors.New("a cannot be nil")
+	default:
+		return nil, errors.New("unknown Authenticator type")
```

Comment:
```suggestion
		return nil, errors.Errorf("unknown Authenticator type: %T", a)
```

Ideally, `case nil` could be removed IMO.